File: Cloud/AzureImageProcessor/WorkerRole.vb
Imports System
Imports System.Collections.Generic
Imports System.Diagnostics
Imports System.Net
Imports System.Threading
Imports Microsoft.WindowsAzure
Imports Microsoft.WindowsAzure.Diagnostics
Imports Microsoft.WindowsAzure.ServiceRuntime
Imports CloudDAL
Imports InternalStorageBrokerServices

Public Class WorkerRole
    Inherits RoleEntryPoint

    Public Overrides Sub Run()

        While (True)

            ' NEW SLICE REQUEST
            Dim requestMsg As SliceRequestMessage =
                If(QueueAccessor.IsAvailable, QueueAccessor.SliceRequestQueue.AcceptMessage(Of SliceRequestMessage)(), Nothing)
            If requestMsg IsNot Nothing Then

                ' check for possible poison message
                If requestMsg.RawMessage.DequeueCount <= requestMsg.Queue.PoisonThreshold Then

                    Try

                        ' instantiate storage accessors
                        Dim imgLibraryAccessor As BlobAccessor = New BlobAccessor(StorageBroker.GetStorageConnectionStringForAccount(requestMsg.ImageLibraryUri))
                        Dim blobAccessor As BlobAccessor = New BlobAccessor(StorageBroker.GetStorageConnectionStringForAccount(requestMsg.SliceUri))
                        Dim tableAccessor As TableAccessor = New TableAccessor(StorageBroker.GetStorageConnectionStringForClient(requestMsg.ClientId))

                        ' use caching? (this is a role property for experimentation, in production you'd likely commit to caching)
                        Dim cacheType As CachingMechanism
                        Try
                            cacheType = CType([Enum].Parse(cacheType.GetType(), RoleEnvironment.GetConfigurationSettingValue("CachingMechanism")), CachingMechanism)
                        Catch e As Exception
                            cacheType = CachingMechanism.InRole
                        End Try

                        ' time the operation
                        Dim sw As New Stopwatch()
                        sw.Start()

                        ' build the library
                        Dim library As New ImageLibrary(imgLibraryAccessor,
                                                        requestMsg.ImageLibraryUri,
                                                        requestMsg.TileSize,
                                                        cacheType)

                        sw.Stop()
                        tableAccessor.WriteStatusEntry(requestMsg.RequestId,
                                 String.Format("Slice {0}: CacheType: {1}, RoleId: {2}",
                                               requestMsg.SliceSequence, cacheType.ToString(), RoleEnvironment.CurrentRoleInstance.Id))
                        sw.Start()

                        ' generate the image
                        Dim generatedImage = library.Mosaicize(
                            blobAccessor.RetrieveImage(requestMsg.SliceUri),
                            requestMsg.TileSize,
                            Sub(msg As String)
                                tableAccessor.WriteStatusEntry(requestMsg.RequestId,
                                            String.Format("Slice {0}: {1}", requestMsg.SliceSequence, msg))
                            End Sub
                        )

                        ' stop the timer
                        sw.Stop()

                        tableAccessor.WriteStatusEntry(requestMsg.RequestId,
                                 String.Format("Slice {0} completed in {1}:{2:00}.{3:000} via RoleId: {4}",
                                               requestMsg.SliceSequence, sw.Elapsed.Minutes, sw.Elapsed.Seconds, sw.ElapsedMilliseconds,
                                               RoleEnvironment.CurrentRoleInstance.Id))
                        tableAccessor.WriteStatusEntry(requestMsg.RequestId,
                                 String.Format("Slice {0} generated Image hits: {1}/{2} ColorValue hits: {3}/{4}",
                                               requestMsg.SliceSequence, library.ImageRetrieves, library.ImageRequests,
                                               library.ColorValueRetrieves, library.ColorValueRequests))

                        ' store the processed segment
                        Dim processedSliceUri = blobAccessor.StoreImage(generatedImage, blobAccessor.SliceOutputContainer,
                                                                        String.Format("{0}_{1:000}", requestMsg.BlobName, requestMsg.SliceSequence))

                        ' signal completion of the slice
                        QueueAccessor.SliceResponseQueue.SubmitMessage(Of SliceResponseMessage)(requestMsg.ClientId, requestMsg.RequestId, processedSliceUri)

                        ' message has been processed
                        requestMsg.DeleteFromQueue()

                    Catch e As Exception

                        Dim tableAccessor As TableAccessor = New TableAccessor(StorageBroker.GetStorageConnectionStringForClient(requestMsg.ClientId))
                        tableAccessor.WriteStatusEntry(requestMsg.RequestId,
                            String.Format("Slice {0}: An unrecoverable exception has occurred, consult diagnostic logs for details.",
                                          requestMsg.SliceSequence))

                        ' something really bad happened!
                        Trace.TraceError("Request: {0}{4}Exception: {1}{4}Message: {2}{4}Trace: {3}",
                            requestMsg.RequestId,
                            e.GetType(),
                            e.Message,
                            e.StackTrace,
                            Environment.NewLine)
                    End Try
                Else

                        ' remove poison message from the queue
                        requestMsg.DeleteFromQueue()
                        Trace.TraceWarning("Request: {0}{4}Exception: {1}{4}Message: {2}{4}Trace: {3}",
                                requestMsg.RequestId,
                                "N/A",
                                String.Format("Possible poison message [id: {0}] removed from queue '{1}' after {2} attempts",
                                    requestMsg.RawMessage.Id,
                                    QueueAccessor.SliceRequestQueue,
                                    QueueAccessor.SliceRequestQueue.PoisonThreshold),
                                "N/A",
                                Environment.NewLine)
                End If
            Else
                Thread.Sleep(10000)
            End If
        End While
    End Sub

    Public Overrides Function OnStart() As Boolean

        ' Set the maximum number of concurrent connections 
        ServicePointManager.DefaultConnectionLimit = 12

        ' For information on handling configuration changes
        ' see the MSDN topic at http://go.microsoft.com/fwlink/?LinkId=166357.

        Return MyBase.OnStart()
    End Function
End Class

File: Cloud/AzureJobController/WorkerRole.cs
using System;
using System.Diagnostics;
using System.Drawing;
using System.Linq;
using System.Net;
using System.Threading;
using CloudDAL;
using InternalStorageBrokerServices;
using Microsoft.WindowsAzure.ServiceRuntime;
using NotificationServices;
using Utilities;

namespace AzureJobController
{
    public class WorkerRole : RoleEntryPoint
    {
        public override void Run()
        {
            TableAccessor tableAccessor = null;
            BlobAccessor blobAccessor = null;

            while (true)
            {
                #region NEW IMAGE REQUEST
                var requestMsg = QueueAccessor.IsAvailable ? QueueAccessor.ImageRequestQueue.AcceptMessage<ImageRequestMessage>() : null;
                if (requestMsg != null)
                {
                    // check for possible poison message
                    if (requestMsg.RawMessage.DequeueCount <= requestMsg.Queue.PoisonThreshold)
                    {
                        try
                        {
                            // instantiate storage accessors
                            blobAccessor = new BlobAccessor(StorageBroker.GetStorageConnectionStringForAccount(requestMsg.ImageUri));
                            tableAccessor = new TableAccessor(StorageBroker.GetStorageConnectionStringForClient(requestMsg.ClientId));

                            // retrieve source image from blob storage
                            var imageBytes = blobAccessor.RetrieveImage(requestMsg.ImageUri);

                            // actual slices may be less than request if image is too small
                            var slices = ImageUtilities.SliceAndDice(imageBytes, requestMsg.Slices).ToList();
                            Byte totalSlices = (Byte)slices.Count;
                            if (requestMsg.Slices != totalSlices)
                            {
                                var job = tableAccessor.GetJobByRequestId(requestMsg.RequestId);
                                if (job != null)
                                {
                                    job.Slices = totalSlices;
                                    tableAccessor.UpdateJob(job);
                                }
                            }

                            // loop through each slice, storing it and submitting a message to queue
                            for (Byte sliceNum = 0; sliceNum < totalSlices; sliceNum++)
                            {
                                Uri slicePath = blobAccessor.StoreImage(
                                    slices[sliceNum], BlobAccessor.SliceInputContainer, String.Format("{0}_{1:000}", requestMsg.BlobName, sliceNum));

                                QueueAccessor.SliceRequestQueue.SubmitMessage<SliceRequestMessage>
                                    (requestMsg.ClientId, requestMsg.RequestId, requestMsg.ImageLibraryUri, slicePath, sliceNum, requestMsg.TileSize);

                                tableAccessor.WriteStatusEntry(requestMsg.RequestId,
                                    String.Format("Dispatched slice {0} of {1}", sliceNum, totalSlices - 1));
                            }

                            // message processed successfully
                            requestMsg.DeleteFromQueue();
                        }
                        catch (Exception e)
                        {
                            // something really bad happened!
                            Trace.TraceError("Request: {0}{4}Exception: {1}{4}Message: {2}{4}Trace: {3}",
                                requestMsg.RequestId,
                                e.GetType(),
                                e.Message,
                                e.StackTrace,
                                Environment.NewLine);
                        }
                    }
                    else
                    {
                        // remove poison message from the queue
                        requestMsg.DeleteFromQueue();
                        Trace.TraceWarning("Request: {0}{4}Exception: {1}{4}Message: {2}{4}Trace: {3}",
                                requestMsg.RequestId,
                                "N/A",
                                String.Format("Possible poison message [id: {0}] removed from queue '{1}' after {2} attempts", 
                                    requestMsg.RawMessage.Id,
                                    QueueAccessor.ImageRequestQueue,
                                    QueueAccessor.ImageRequestQueue.PoisonThreshold),
                                "N/A",
                                Environment.NewLine);
                    }
                }
                #endregion

                #region COMPLETED SLICE NOTIFICATION
                var responseMsg = QueueAccessor.IsAvailable ? QueueAccessor.SliceResponseQueue.AcceptMessage<SliceResponseMessage>() : null;
                if (responseMsg != null)
                {
                    // check for possible poison message
                    if (responseMsg.RawMessage.DequeueCount <= responseMsg.Queue.PoisonThreshold)
                    {
                        try
                        {
                            // instantiate storage accessors
                            blobAccessor = new BlobAccessor(StorageBroker.GetStorageConnectionStringForAccount(responseMsg.SliceUri));
                            tableAccessor = new TableAccessor(StorageBroker.GetStorageConnectionStringForClient(responseMsg.ClientId));

                            // get the job data (from Azure table) indicating number of slices and the tile size
                            var job = tableAccessor.GetJobByRequestId(responseMsg.RequestId);
                            if ((job != null) && (job.Slices > 0) && (job.TileSize > 0) && (job.SourceWidth > 0) && (job.SourceHeight > 0))
                            {
                                // get slices processed so far
                                var processedSliceUris = blobAccessor.RetrieveImageUris(BlobAccessor.SliceOutputContainer, responseMsg.BlobName).ToList();

                                // if we have the expected number, this is the last slice to complete
                                if (processedSliceUris.Count() == (Int32)job.Slices)
                                {
                                    Byte[] finalImgBytes = ImageUtilities.StitchAndMend(
                                        processedSliceUris, blobAccessor.RetrieveImages,
                                        (byte)job.TileSize, new Size(job.SourceWidth, job.SourceHeight));

                                    Uri finalImageUri = blobAccessor.StoreImage(finalImgBytes, BlobAccessor.ImageOutputContainer, responseMsg.BlobName);

                                    QueueAccessor.ImageResponseQueue.SubmitMessage<ImageResponseMessage>
                                        (responseMsg.ClientId, responseMsg.RequestId, finalImageUri);
                                }
                            }
                            else
                            {
                                throw new SystemException("Metadata for job {0} in Jobs table was not found or is invalid");
                            }

                            // message processed successfully
                            responseMsg.DeleteFromQueue();
                        }
                        catch (Exception e)
                        {

                            // something really bad happened!
                            Trace.TraceError("Request: {0}{4}Exception: {1}{4}Message: {2}{4}Trace: {3}",
                                responseMsg.RequestId,
                                e.GetType(),
                                e.Message,
                                e.StackTrace,
                                Environment.NewLine);
                        }
                    }
                    else
                    {
                        // remove poison message from the queue
                        responseMsg.DeleteFromQueue();
                        Trace.TraceWarning("Request: {0}{4}Exception: {1}{4}Message: {2}{4}Trace: {3}",
                                responseMsg.RequestId,
                                "N/A",
                                String.Format("Possible poison message [id: {0}] removed from queue '{1}' after {2} attempts", 
                                    responseMsg.RawMessage.Id,
                                    QueueAccessor.SliceResponseQueue,
                                    QueueAccessor.SliceResponseQueue.PoisonThreshold),
                                "N/A",
                                Environment.NewLine);
                    }
                }
                #endregion

                #region MOST SINCERELY DONE NOTIFICATION
                var completeMsg = QueueAccessor.IsAvailable ? QueueAccessor.ImageResponseQueue.AcceptMessage<ImageResponseMessage>() : null;
                if (completeMsg != null) 
                {
                    // check for possible poison message
                    if (completeMsg.RawMessage.DequeueCount <= completeMsg.Queue.PoisonThreshold)
                    {
                        try
                        {
                            // instantiate storage accessors
                            blobAccessor = new BlobAccessor(StorageBroker.GetStorageConnectionStringForAccount(completeMsg.ImageUri));
                            tableAccessor = new TableAccessor(StorageBroker.GetStorageConnectionStringForClient(completeMsg.ClientId));

                            // get the completed job entity from table storage
                            var job = tableAccessor.GetJobByRequestId(completeMsg.RequestId);
                            if (job != null)
                            {

                                // update pointer to completed image (and the finish time)
                                job.FinalImageUri = completeMsg.ImageUri.ToString();
                                job.FinishTime = DateTime.UtcNow;

                                tableAccessor.UpdateJob(job);

                                tableAccessor.WriteStatusEntry(completeMsg.RequestId, "Request complete.");
                            }
                            else
                            {
                                Trace.TraceWarning("Request: {0}{4}Exception: {1}{4}Message: {2}{4}Trace: {3}",
                                    completeMsg.RequestId,
                                    "N/A",
                                    "Request completed, but record not found in Jobs table",
                                    "N/A",
                                    Environment.NewLine);
                            }

                            // only notify user (or try to) once!
                            if (completeMsg.RawMessage.DequeueCount == 1)
                            {
                                try
                                {
                                    // notify via service bus that request is complete
                                    if (job != null)
                                    {
                                        Notifier.SendRequestCompletedNotification(
                                            job.ClientId,
                                            completeMsg.RequestId,
                                            completeMsg.ImageUri);
                                    }
                                }
                                catch (Exception e)
                                {
                                    Trace.TraceWarning("Request: {0}{4}Exception: {1}{4}Message: {2}{4}Trace: {3}",
                                        completeMsg.RequestId,
                                        e.GetType(),
                                        e.Message,
                                        e.StackTrace,
                                        Environment.NewLine);
                                }
                            }

                            // message processed successfully
                            completeMsg.DeleteFromQueue();
                        }

                        catch (Exception e)
                        {
                            // something really bad happened!
                            Trace.TraceError("Request: {0}{4}Exception: {1}{4}Message: {2}{4}Trace: {3}",
                                completeMsg.RequestId,
                                e.GetType(),
                                e.Message,
                                e.StackTrace,
                                Environment.NewLine);
                        }
                    }
                    else
                    {
                        // remove poison m essage from the queue
                        completeMsg.DeleteFromQueue();
                        Trace.TraceWarning("Request: {0}{4}Exception: {1}{4}Message: {2}{4}Trace: {3}",
                                completeMsg.RequestId,
                                "N/A",
                                String.Format("Possible poison message [id: {0}] removed from queue '{1}' after {2} attempts", 
                                    completeMsg.RawMessage.Id,
                                    QueueAccessor.ImageResponseQueue,
                                    QueueAccessor.ImageResponseQueue.PoisonThreshold),
                                "N/A",
                                Environment.NewLine);
                    }
                }
                #endregion

                // only sleep if we didn't do anything this time through
                if ((requestMsg == null) && (responseMsg == null) && (completeMsg == null))
                {
                    Thread.Sleep(10000);
                }
            }
        }

        public override bool OnStart()
        {
            // Set the maximum number of concurrent connections 
            ServicePointManager.DefaultConnectionLimit = 12;

            // For information on handling configuration changes
            // see the MSDN topic at http://go.microsoft.com/fwlink/?LinkId=166357.

            return base.OnStart();
        }
    }
}